Initial Synopsys Operator commit

davemeurer opened the pull request:
The Synopsys Operator contains the ability to deploy Synopsys Black Duck and the OpsSight connector, which automatically performs an OSS scan on all container images deployed in an OpenShift cluster.

SamiSousa commented on community-operators/synopsys/synopsys-operator.2019.1.0.yaml:
```diff
@@ -0,0 +1,288 @@
+metadata:
+  name: synopsys-operator.2019.1.0
+  namespace: placeholder
+  annotations:
```
Additional annotations can be found here: https://github.com/operator-framework/community-operators/blob/master/docs/required-fields.md

davemeurer replied:
fixed in latest push

SamiSousa commented on community-operators/synopsys/synopsys-operator.2019.1.0.yaml:
```diff
@@ -0,0 +1,288 @@
+  name: synopsys-operator.2019.1.0
+  namespace: placeholder
+  annotations:
+      categories: "Application Security, Monitoring"
```
Be sure your categories match values given here: https://github.com/operator-framework/community-operators/blob/master/docs/required-fields.md#categories

davemeurer replied:
fixed in latest push

SamiSousa commented on community-operators/synopsys/synopsys-operator.2019.1.0.yaml:
OwnNamespace should be set to true if your operator watches all namespaces.
@alecmerdler

davemeurer replied:
fixed in latest push

SamiSousa commented on community-operators/synopsys/synopsys-operator.2019.1.0.yaml:
For nonResourceUrls, no need to include null keys

davemeurer replied:
should have this fixed tomorrow morning

davemeurer replied:
should be fixed now

SamiSousa commented on community-operators/synopsys/synopsys-operator.2019.1.0.yaml:
The permissions on this operator are too broad, please reduce them to what the operator needs to perform its function

davemeurer replied:
should have this fixed tomorrow morning

davemeurer replied:
new permissions block added, please let me know if that works

davemeurer commented:
Thanks for all the great feedback @SamiSousa .. I hope to make the changes and test in OpenShift today/tomorrow.

davemeurer commented on community-operators/synopsys/synopsys-operator.package.yaml:
```diff
@@ -0,0 +1,4 @@
+packageName: synopsys-operator
```
fixed in latest push

davemeurer commented on community-operators/synopsys/synopsys-operator.package.yaml:
```diff
@@ -0,0 +1,4 @@
+packageName: synopsys-operator
+channels:
+    - name: stable
```
fixed in latest push

davemeurer commented:
I pushed a couple changes up.  Still need to work out the permissions and test the operator on OpenShift.  Got a bit stuck creating an OperatorSource CR, but working thru it.

davemeurer commented:
Not sure why my builds are failing.  Any thoughts?  The travis logs mention something about the wrong version format on the CSV?  I moved the files and renamed, so not sure if that is causing issues?

SamiSousa commented:
@davemeurer We are currently testing the CI tests. You are correct in seeing that there's an issue in the assumed format of some values in the operator bundle. I'm working on a fix at the moment.

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,321 @@
+      createdAt: 2018-12-14T17:09:00Z
+      support: Synopsys
+      certified: "False"
+      alm-examples: '[{"apiVersion": "synopsys.com/v1", "kind": "OpsSight", "metadata": { "name": "opssight-test" }, "spec": { "namespace": "opssight-test", "perceptor": { "name": "opssight-core", "port": 3001, "image": "docker.io/blackducksoftware/opssight-core:2.2.2", "checkForStalledScansPauseHours": 999999, "stalledScanClientTimeoutHours": 999999, "modelMetricsPauseSeconds": 15, "unknownImage ...
```
Is there an example for the Alerts CR? Or is this CR used internally by the operator?

davemeurer replied:
Good question.  It's an optional CR, and at the moment we are planning to not include it.  I'll remove the alerts.crd and commit to restart the CI tests.

davemeurer replied:
We added it back in, and an example exists now.

SamiSousa commented:
@davemeurer Could you push an empty "dummy" commit? This is needed to restart the CI tests on this PR. Apologies for the inconvenience

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,346 @@
+      support: Synopsys
+      certified: "False"
+      alm-examples: '[{"apiVersion": "synopsys.com/v1", "kind": "OpsSight", "metadata": { "name": "opssight-test" }, "spec": { "namespace": "opssight-test", "perceptor": { "name": "opssight-core", "port": 3001, "image": "docker.io/blackducksoftware/opssight-core:2.2.2", "checkForStalledScansPauseHours": 999999, "stalledScanClientTimeoutHours": 999999, "modelMetricsPauseSeconds": 15, "unknownImage ...
+      { "apiVersion": "synopsys.com/v1", "kind": "Blackduck", "metadata": { "name": "blackduck-ephemeral" }, "spec": { "namespace": "blackduck-ephemeral", "licenseKey": "<BLACKDUCK_LICENSE_KEY>", "certificateName": "default", "environs": [ "HTTPS_VERIFY_CERTS:yes", "HUB_POSTGRES_ENABLE_SSL:false", "HUB_VERSION:2018.12.2", "IPV4_ONLY:0", "RABBITMQ_DEFAULT_VHOST:protecodesc", "USE_ALERT:0", "CFSSL: ...
```
I'm not sure this is valid YAML. This should be indented further to indicate it's part of the string value for `alm-examples`

SamiSousa replied:
Actually I think this is getting picked up, but I still would prefer the formatting be unambiguous

davemeurer replied:
Done locally, will commit with others below.

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.1.0.clusterserviceversion.yaml:
OLM currently only supports `deployments` for operators. Is there a reason why a ReplicationController is preferred?

msenmurugan replied:
@SamiSousa since we are supporting customers using openshift 3.6 and 3.7, I don't think deployments are available in those version.

davemeurer replied:
updated locally. We used replicationcontrollers for 3.6 compatibility, but we can doc that exception since most users will be on 4.0 using these Operator files

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.1.0.clusterserviceversion.yaml:
OLM doesn't support `replicationcontrollers`

msenmurugan replied:
same comment as above

davemeurer replied:
same comment as above. Thanks for pointing this out!!

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.1.0.clusterserviceversion.yaml:
This should be a list

davemeurer replied:
done.  about to commit another round of changes...

SamiSousa commented:
@davemeurer OLM doesn't support replicationcontrollers, which is why CI is failing.

jdob commented on upstream-community-operators/synopsys/blackducks.crd.yaml:
```diff
@@ -0,0 +1,55 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: 2019-03-04T15:38:44Z
```
The CRDs shouldn't have the deployment info in them, so things like the creationTimestamp, link, version, etc. should all be removed from the actual definition and left for OpenShift to fill out when it is created.

jdob commented on upstream-community-operators/synopsys/blackducks.crd.yaml:
```diff
@@ -0,0 +1,55 @@
+  - name: v1
+    served: true
+    storage: true
+status:
```
Similarly, the entire status section should be removed, since that only applies to created resources.

jdob commented on upstream-community-operators/synopsys/blackducks.crd.yaml:
```diff
@@ -0,0 +1,55 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: 2019-03-04T15:38:44Z
```
The CRDs shouldn't have the deployment info in them, so things like the creationTimestamp, link, version, etc. should all be removed from the actual definition and left for OpenShift to fill out when it is created.

davemeurer replied:
We removed those in the latest commit.  Thanks @jdob!

jdob commented on upstream-community-operators/synopsys/blackducks.crd.yaml:
```diff
@@ -0,0 +1,55 @@
+  - name: v1
+    served: true
+    storage: true
+status:
```
Similarly, the entire status section should be removed, since that only applies to created resources.

davemeurer replied:
Status was removed as well in the latest commit.

SamiSousa commented:
@davemeurer The CSV has been renamed, but I believe my comments still stand. Please address my comments as the CSV will not work with OLM otherwise

davemeurer commented:
@SamiSousa Thanks for the feedback.  I will address your comments.

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.4.0.clusterserviceversion.yaml:
Please use lowercase `deployment` for `spec.install.strategy`

davemeurer replied:
Done

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.4.0.clusterserviceversion.yaml:
Is there any documentation for this `SEAL_KEY`? Could this be added to the description if this is required for the operator to run?

davemeurer replied:
Yes, documentation will be soon published in our external docs site, which is referenced in the links section.

SamiSousa commented on upstream-community-operators/synopsys/synopsys-operator.2019.4.0.clusterserviceversion.yaml:
I believe this is misaligned. Please confirm that your deployment is valid

davemeurer replied:
apologies, you are correct. fixed locally.

SamiSousa commented:
@davemeurer The CI logs show that the CSV fails to install due to the following error: https://travis-ci.com/operator-framework/community-operators/builds/109351842#L1404
To summarize: `'install strategy failed: Deployment.apps "synopsys-operator" is invalid:
      [spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string(nil),
      MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: empty selector is invalid
      for deployment, spec.template.spec.containers[0].volumeMounts[0].name: Not found:
      "synopsys-operator", spec.template.spec.containers[0].volumeMounts[1].name:
      Not found: "synopsys-operator-tls"]'`

msenmurugan commented:
@SamiSousa Looks like the CSV file is good and CI build is good till the end but we don't know the root cause of the build failure. Can you please look into it?

SamiSousa commented:
@msenmurugan No need for your operator to be written with the SDK to be installed via OLM :slightly_smiling_face: 
Does your operator need these configmaps and secrets to exist in order to run?

SamiSousa commented:
@davemeurer @msenmurugan Has your team tested these updated manifests against OLM on OpenShift and on vanilla Kubernetes?

davemeurer commented:
Hi @SamiSousa, @msenmurugan has made some changes, and the build is failing to pick up the subscription.  Any thoughts on that error?

davemeurer commented:
Hi @SamiSousa - we finally found the issue with our CSV.  The build has passed... yay!!

davemeurer commented:
Also, to answer your previous question - we have tested this operator on OpenShift 4.0

SamiSousa commented:
@davemeurer Please rebase your commits so that I can add this operator

davemeurer commented:
Thanks @SamiSousa - I just did the rebase

SamiSousa commented:
@davemeurer Still looks like there are 37 commits for this operator; please rebase it to reduce the number of commits

davemeurer commented:
rebased @SamiSousa - yup, sorry we made some tweaks last night and I didn't rebase again.  It should be all set.  Appreciate all your help and patience!

SamiSousa commented:
@davemeurer Now I see 107 commits, with changes to 31 files. Please rebase your commits, if you need help I'd be happy to provide guidance

davemeurer commented:
Thanks Sami, I'm doing:
$ git fetch upstream
$ git rebase upstream/master

Is that correct? (sorry, I'm not a git rebase expert)

SamiSousa commented:
Try creating a test branch, setting it to the latest upstream master:
```
$ git checkout -b test
$ git fetch upstream master
$ git reset --hard upstream/master
```
Then apply your PR branch commits on top (I believe you are using master locally, but change accordingly to your case):
```
$ git merge --squash master
$ git commit -m "[upstream] Synopsys Operator"
```
Then push this to your remote branch:
```
$ git push -f origin test:master
```
This is a quick way of getting the changes sorted out, let me know if you have any trouble with these commands.

davemeurer commented:
Thank you for the assistance @SamiSousa, I've followed your suggestions and force pushed my test branch.

SamiSousa commented:
@davemeurer Thank you for the PR!
/lgtm

davemeurer commented:
Hi @SamiSousa - there are actually no changes with this latest push.  I was just syncing my local master branch.

SamiSousa commented:
/lgtm

davemeurer commented:
Thanks @SamiSousa - what would be the next steps for us getting this published on operatorhub.io?
